<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Pandora: README</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Pandora
   &#160;<span id="projectnumber">Graph based genome analysis.</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">README </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a href="https://travis-ci.org/rmcolq/pandora"></a> master</p>
<p><a href="https://travis-ci.com/rmcolq/pandora"></a> dev</p>
<p><a href="https://singularity-hub.org/collections/1285"></a></p>
<h1>pandora</h1>
<h3>Note this is in active rapid development at present and not ready for reliable use</h3>
<p>Pandora is a tool for bacterial genome analysis without using a reference genome, including genetic variation from SNPs to gene presence/absence across the whole pan-genome. Core ideas are:</p><ul>
<li>new samples look like recombinants (plus mutations) of things seen before</li>
<li>we should be analysing nucleotide-level variation everywhere, not just in core genes</li>
<li>arbitrary reference genomes are unnatural</li>
</ul>
<p>Pandora works with Illumina or nanopore data, allowing per-sample analysis (sequence inference and SNP/indel/gene-calling) and comparison of multiple samples. To do this it uses population reference graphs (PRG) which have been built for orthologous blocks of interest (e.g. genes and intergenic regions). See <a href="https://github.com/rmcolq/make_prg">https://github.com/rmcolq/make_prg</a> for a pipeline which can construct these PRGs from a set of aligned sequence files.</p>
<p>It can do the following for a single sample (read dataset):</p>
<ul>
<li>Output inferred gene sequences for the orthologous chunks (eg genes) in the PRG</li>
<li>Output a VCF showing the variation found in the pangenome genes which are present, with respect to any reference in the PRG.</li>
</ul>
<p>For a collection of samples, it can:</p>
<ul>
<li>Output a matrix showing inferred copy-number of each gene in each sample genome.</li>
<li>Output one VCF per orthologous-chunk, showing how samples which contained this chunk differed in their gene sequence. Variation is shown with respect to the most informative recombinant path in the PRG . Soon, in a galaxy not so far away, it will allow<ul>
<li>discovery of new variation not in the PRG</li>
</ul>
</li>
</ul>
<p>Warning - this code is still in development.</p>
<h2>Installation</h2>
<ul>
<li>Requires a Unix or Mac OS.</li>
<li>Requires a system install of <code>zlib</code>. If this is not already installed, <a href="https://geeksww.com/tutorials/libraries/zlib/installation/installing_zlib_on_ubuntu_linux.php">this</a> tutorial is helpful.</li>
<li>Requires a system installation of <code>boost</code> containing the <code>system</code>, <code>filesystem</code>, <code>log</code> (which also depends on <code>thread</code> and <code>date_time</code>) and <code>iostreams</code> libraries. If not already installed use the following or look at <a href="https://www.boost.org/doc/libs/1_62_0/more/getting_started/unix-variants.html">this</a> guide. <pre class="fragment">wget https://sourceforge.net/projects/boost/files/boost/1.62.0/boost_1_62_0.tar.gz --no-check-certificate
tar xzf boost_1_62_0.tar.gz
cd boost_1_62_0
./bootstrap.sh [--prefix=/prefix/path] --with-libraries=system,filesystem,iostreams,log,thread,date_time
./b2 install
</pre></li>
<li>Download and install <code>pandora</code> as follows: <pre class="fragment">git clone https://github.com/rmcolq/pandora.git
cd pandora
mkdir build
cd build
cmake [-DCMAKE_PREFIX_PATH=/prefix/path] ..
make
ctest -VV
cd ..
</pre></li>
</ul>
<h2>Singularity Container</h2>
<p>Instead you can download and use the singularity container: </p><pre class="fragment">singularity pull --force --name pandora.simg shub://rmcolq/pandora:pandora
singularity exec pandora.simg pandora
</pre><h2>Usage</h2>
<h3>Population Reference Graphs</h3>
<p>Pandora assumes you have already constructed a fasta-like file of graphs, one entry for each gene/ genome region of interest.</p>
<h3>Build index</h3>
<p>Takes a fasta-like file of PRG sequences and constructs an index, and directory of gfa files to be used by pandora map. </p><pre class="fragment">  Usage: pandora index [options] &lt;prgs.fa&gt;
  Options:
    -h,--help           Show this help message
    -w W                Window size for (w,k)-minimizers, default 14
    -k K                K-mer size for (w,k)-minimizers, default 15
</pre><p>The index stores (w,k)-minimizers for each PRG path found. These parameters can be specified, but default to w=1, k=15.</p>
<h3>Map reads to index</h3>
<p>This takes a fasta of noisy long read sequence data and compares to the index. It infers which of the PRG genes/elements is present, and for those that are present it outputs the inferred sequence. </p><pre class="fragment">  Usage: pandora map -p PRG_FILE -r READ_FILE -o OUTDIR &lt;option(s)&gt;
  Options:
   -h,--help             Show this help message
   -p,--prg_file PRG_FILE    Specify a fasta-style prg file
   -r,--read_file READ_FILE  Specify a file of reads in fasta format
   -o,--outdir OUTDIR            Specify directory of output
   -w W              Window size for (w,k)-minimizers, must be &lt;=k, default 14
   -k K              K-mer size for (w,k)-minimizers, default 15
   -m,--max_diff INT         Maximum distance between consecutive hits within a cluster, default 500 (bps)
   -e,--error_rate FLOAT     Estimated error rate for reads, default 0.11
   --genome_size NUM_BP          Estimated length of genome, used for coverage estimation
   --output_kg           Save kmer graphs with fwd and rev coverage annotations for found localPRGs
   --output_vcf          Save a vcf file for each found localPRG
   --vcf_refs REF_FASTA      A fasta file with an entry for each LocalPRG giving reference sequence for
                                 VCF. Must have a perfect match in the graph and the same name as the graph
   --illumina            Data is from illumina rather than nanopore, so is shorter with low error rate
   --bin             Use binomial model for kmer coverages, default is negative binomial
   --max_covg            Maximum average coverage from reads to accept
   --regenotype          Add extra step to carefully genotype SNP sites</pre> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
